flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierUnaligner.java

```diff
@@ -91,31 +91,25 @@
 	private final ThreadSafeUnaligner threadSafeUnaligner;
 
 	CheckpointBarrierUnaligner(
-			int[] numberOfInputChannelsPerGate,
 			SubtaskCheckpointCoordinator checkpointCoordinator,
 			String taskName,
-			AbstractInvokable toNotifyOnCheckpoint) {
+			AbstractInvokable toNotifyOnCheckpoint,
+			IndexedInputGate... inputGates) {
 		super(toNotifyOnCheckpoint);
 
 		this.taskName = taskName;
-
-		final int numGates = numberOfInputChannelsPerGate.length;
-
-		gateChannelOffsets = new int[numGates];
-		for (int index = 1; index < numGates; index++) {
-			gateChannelOffsets[index] = gateChannelOffsets[index - 1] + numberOfInputChannelsPerGate[index - 1];
-		}
-
-		final int totalNumChannels = gateChannelOffsets[numGates - 1] + numberOfInputChannelsPerGate[numGates - 1];
-		hasInflightBuffers = new boolean[totalNumChannels];
-
-		channelInfos = IntStream.range(0, numGates)
-			.mapToObj(gateIndex -> IntStream.range(0, numberOfInputChannelsPerGate[gateIndex])
-				.mapToObj(channelIndex -> new InputChannelInfo(gateIndex, channelIndex)))
-			.flatMap(Function.identity())
+		this.channelInfos = Arrays.stream(inputGates)
+			.flatMap(gate -> gate.getChannels().stream().map(InputChannel::getChannelInfo))
 			.toArray(InputChannelInfo[]::new);
-
-		threadSafeUnaligner = new ThreadSafeUnaligner(totalNumChannels,	checkNotNull(checkpointCoordinator), this);
+		hasInflightBuffers = new boolean[channelInfos.length];
+		threadSafeUnaligner = new ThreadSafeUnaligner(channelInfos.length, checkNotNull(checkpointCoordinator), this);
+
+		gateChannelOffsets = new int[inputGates.length];
+		int offset = 0;
+		for (final IndexedInputGate gate: inputGates) {
+			gateChannelOffsets[gate.getGateIndex()] = offset;
+			offset += gate.getNumberOfInputChannels();
+		}
```

Comment:
I see similar code in `CheckpointBarrierAligner` constructor and same in `InputProcessorUtil.createCheckpointedInputGates`.
WDYT about extracting it into something like `InputGateResolver` and passing it to barrier handlers, input gates and whoever else need it?

Ideally, I think we should use self-sufficient IDs, but I guess this would be too invasive change.

Reply:
That's a tough one; I had feared you pointing it out. The issue is that they are all similar but in the end also quite different.
`CheckpointBarrierAligner` creates a map gate->offset, because it also creates an index->gate array, which it uses in `resumeConsumption` to first resolve the gate for a given flattened channel index and then delegate `resumeConsumption` with adjusted index (that's where the offset is needed).
`CheckpointBarrierUnaligner` mainly uses the `InputChannelInfo` and hence creates a gate index->offset map to create the flattened index for other `BarrierHandler` stuff.
These two versions gets the `SingleInputGates` (only they have indices).

`InputProcessorUtil.createCheckpointedInputGates` uses union input gates, so it cannot use indexes. So it goes by list index. However, now that I write it down this looks also suspicious...
I guess to use simple offsets in `CheckpointedInput` (because of union gates), we actually need to assume consecutive indices... @pnowojski however also saw some strange cases, maybe he can add his observations.

Reply:
I look at it from a bit different perspective. `InputProcessorUtil.createCheckpointedInputGates` passes offsets to `CheckpointedInputGate` which uses them to talk with `barrierHandler`:
```
// CheckpointedInputGate with offsetChannelIndex() inlined for clarity
barrierHandler.isBlocked(bufferOrEvent.getChannelIndex() + channelIndexOffset)
barrierHandler.processBarrier(checkpointBarrier, bufferOrEvent.getChannelIndex() + channelIndexOffset);
```
`CheckpointBarrierAligner.processBarrier` internally calls `resumeConsumption` which unflattens index back:
```
private void resumeConsumption(int channelIndex) {
    InputGate inputGate = channelIndexToInputGate[channelIndex];
    inputGate.resumeConsumption(channelIndex - inputGateToChannelIndexOffset.get(inputGate));
}
```
So, they **are** related.

Now, are they equal?
The fact that `CheckpointedInputGate` can use `UnionInputGate` is accounted by 
`UnionInputGate.getNumberOfInputChannels` which returns the total number of input channels. So these offsets indeed are the same.

I'm more inclined now to use `gateId+channelId` pairs or `InputChannelInfo` s :)

Reply:
I'd actually love to use `InputChannelInfo`s everywhere. Previous attempts to refactor proved difficult however. I'll check again.

Reply:
@AHeise are you asking about `InputProcessorUtil#generateInputGateToChannelIndexOffsetMap` not using the input gate indexes? I'm not sure. I guess it is inconsistent with the `numberOfInputChannelsPerGate` but it might be correct, as all of the references/usages of `inputGateToChannelIndexOffset` are going through the offsetting that also uses `inputGateToChannelIndexOffset` (for example `CheckpointedInputGate#offsetChannelIndex`). But it's really hard to reason about.

It's also quite strange that when polling records from input gate, we are first offsetting them once via using `CheckpointedInputGate#offsetChannelIndex` and then we are using basically the same information to revert the offset back, to the original value in `CheckpointBarrierAligner#resumeConsumption`.

If we could simplify it by always using pair of ids without affecting performance that would be really nice...